### DISCUSSION TITLE
Using watch seems to make webpack infinitely recompile itself over and over

### DISCUSSION THREAD

**Discussion by ghost** [webpack/webpack#13959] [Q&A]
When using:

````
config.watch = true
config.watchOptions = {
  aggregateTimeout: 200,
  poll: 1000,
  followSymlinks: true,
  ignored: /node_modules/,
}
````

Webpack repeatedly recompiles itself over and over after the first file change:

````
webpack compiled successfully
webpack compiled successfully
webpack compiled successfully
webpack compiled successfully
webpack compiled successfully
webpack compiled successfully
webpack compiled successfully
webpack compiled successfully
webpack compiled successfully
````

That happens on every poll, but if you don't specify a poll it recompiles itself every couple of milliseconds.  

I have no idea why but it seems as if webpack compiles -> triggers on its own file change event -> recompiles again, over and over.  

I've removed and uninstalled and uncommented just about everything I have, what's left is basically as little as webpack can have to still function, along with ````ts-loader````, so the only real external variable here is ````ts-loader````, but I have no idea if that's relevant or not.  Perhaps ts compiling files is interfering with webpack's ability to pick up on file changes or something. Disabling watching in ````tsconfig.json```` has no effect on it, still the same.  Trying to ignore any  files using ````ignore```` or the ignore plugin has no affect, just gets "ignored".  

````"webpack": "^5.49.0"````

**Comment by alexander-akait** [comment]
Please create reproducible test repo, this is most likely due to additional plugins

**Comment by ghost** [comment]
I've pinpointed this down to exactly what's causing it.  

````
"composite": true
````

This line in ````tsconfig.json```` along with using ````ts-loader```` will make webpack repeatedly recompile itself on the first file change.  If you just make that false or comment it, everything works as expected.

**Reply by ghost** [reply]
I don't know the detail but according to their documentation "The composite option enforces certain constraints which make it possible for build tools (including TypeScript itself, under --build mode) to quickly determine if a project has been built yet." Something in ````ts-loader```` is causing typescript to think it's not been built yet, then it's communicating that to webpack somehow and webpack thinks it needs to recompile itself, even if it has already been built.

**Reply by alexander-akait** [reply]
Maybe better to open this in `ts-loader`, anyway if you provide example how I can reproduce it I will help to investigate problem place(s)

**Reply by ghost** [reply]
Yes I'll open an issue in ts-loader for it.  However if anyone else stumbles on this in the future, now you know what to tell them  Was really frustrating to figure this out lol.  

Here's the repo I've been working on: https://github.com/matriarx/ui

You can replicate it by just running ````npm run dev```` to let webpack compile in watch mode, after which you can edit the ````src/ui/index.tsx```` file and it starts happening.  If it doesn't happen on the first change, it will happen on the second or third time, so then you can just rerun ````npm run dev```` and make another change, then it starts happening.

**Reply by alexander-akait** [reply]
~~What is your OS?, on linux (Ubuntu) all is fine~~

**Reply by alexander-akait** [reply]
Reproduced, problem with `/types/src/ui/index.d.ts` file

**Reply by alexander-akait** [reply]
Try to use `"include": ["src/**/*.ts", "src/**/*.tsx"],` in `tsconfig.json`, ts preloads these types due the `include` option

**Reply by ghost** [reply]
Okay so that actually fixes the issue, however if I just use include "src*" then the IDE can no longer pick up on any .ts files or types outside of src and it also makes the IDE unable to find any types for any imports like i.e. tests.

**Reply by ghost** [reply]
Adding "types" to "exclude" fixes the issue... What a rabbit hole this was... Thank you very much.  

At least hopefully it helps someone else in the future.

**Reply by alexander-akait** [reply]
I think you need to open an issue in IDE, `types` is generated by ts, so they should be not included in `tsconfig.json`

**Reply by alexander-akait** [reply]
We got watching paths from `ts`, and ask webpack to watch these files, so maybe it can be better documented on ts side

**Reply by ghost** [reply]
Alright thank you.  I've opened an issue on ts side because even though this fixes the issue it still doesn't seem right that the types should be making webpack recompile.
